File: src/grouping.rs
//! Key parsing: split a Redis key into a `namespace` and a normalized `type`.
//!
//! Shopware (and the Symfony cache it wraps) produce keys shaped like:
//!
//! ```text
//! WpDcVyo5fP:product-detail-route-1195e812ffd3266c2ed02a5d603d26f5-481ce4e8cb0b80a09beac0aaeea37457
//! WpDcVyo5fP:\x01tags\x01product-0f483ff8ad7735e5e2dc0f44728d7bdb
//! WpDcVyo5fP:cached-product-df61c92879d105261d7fc1e860f1acc7-742e3c7d1b07ae59066c16c7d3f35d4d-450e84ef6ea42a6d10b09c1b4c84ec89
//! ```
//!
//! The first colon-delimited segment is the *namespace* (a per-install random
//! prefix). What follows identifies the cache *type*; we normalize it by
//! stripping the trailing hash segments so that the thousands of individual
//! product caches collapse into one `product-detail-route` bucket.

/// Result of parsing a single key.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParsedKey {
    pub namespace: String,
    pub key_type: String,
}

/// Returns true if `s` looks like a hex hash segment (md5/sha-ish): all hex
/// digits and reasonably long. We require length >= 16 to avoid eating short
/// meaningful words that happen to be hex (e.g. "face", "dead").
fn is_hash_segment(s: &str) -> bool {
    s.len() >= 16 && s.bytes().all(|b| b.is_ascii_hexdigit())
}

/// Split a raw key into namespace + normalized type.
///
/// `namespace_len` is how many leading bytes of the key are treated as the
/// namespace when there is no colon (mirrors the PHP "first N chars" fallback).
pub fn parse_key(key: &str, namespace_len: usize) -> ParsedKey {
    // The Symfony tag marker is the byte 0x01. Render it as the readable token
    // "\x01" (already-escaped input passes through unchanged, since it has no
    // literal 0x01 byte to replace).
    let normalized = key.replace('\u{0001}', "\\x01");

    // Namespace = text before the first ':'. Fall back to first N chars.
    let (namespace, body) = match normalized.find(':') {
        Some(idx) => (normalized[..idx].to_string(), &normalized[idx + 1..]),
        None => {
            let n = namespace_len.min(normalized.len());
            (normalized[..n].to_string(), "")
        }
    };

    let key_type = normalize_type(body);
    ParsedKey {
        namespace,
        key_type,
    }
}

/// Strip trailing hash segments from the key body to derive a stable type name.
///
/// `product-detail-route-<hash>-<hash>` -> `product-detail-route`
/// `\x01tags\x01product-<hash>`         -> `\x01tags\x01product`
fn normalize_type(body: &str) -> String {
    if body.is_empty() {
        return "(no-type)".to_string();
    }

    // Preserve the \x01tags\x01 prefix verbatim, normalize only the remainder.
    const TAG_PREFIX: &str = "\\x01tags\\x01";
    let (prefix, rest) = if let Some(stripped) = body.strip_prefix(TAG_PREFIX) {
        (TAG_PREFIX, stripped)
    } else {
        ("", body)
    };

    let mut segments: Vec<&str> = rest.split('-').collect();

    // Pop trailing hash-looking segments, but always keep at least the first
    // segment so we never collapse everything to an empty string.
    while segments.len() > 1 && is_hash_segment(segments.last().unwrap()) {
        segments.pop();
    }

    // Edge case: a body that is *only* a hash (no descriptive prefix) — keep a
    // generic bucket rather than the raw hash, which would explode cardinality.
    if segments.len() == 1 && is_hash_segment(segments[0]) {
        return format!("{prefix}(hash)");
    }

    format!("{prefix}{}", segments.join("-"))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn t(key: &str) -> ParsedKey {
        parse_key(key, 10)
    }

    #[test]
    fn splits_namespace_and_strips_double_hash() {
        let p = t("WpDcVyo5fP:product-detail-route-1195e812ffd3266c2ed02a5d603d26f5-481ce4e8cb0b80a09beac0aaeea37457");
        assert_eq!(p.namespace, "WpDcVyo5fP");
        assert_eq!(p.key_type, "product-detail-route");
    }

    #[test]
    fn strips_single_hash() {
        let p = t("WpDcVyo5fP:faq-group-930338b2525f792b15a15f598e3ade05");
        assert_eq!(p.key_type, "faq-group");
    }

    #[test]
    fn handles_triple_hash() {
        let p = t("WpDcVyo5fP:cached-product-df61c92879d105261d7fc1e860f1acc7-742e3c7d1b07ae59066c16c7d3f35d4d-450e84ef6ea42a6d10b09c1b4c84ec89");
        assert_eq!(p.key_type, "cached-product");
    }

    #[test]
    fn preserves_tag_prefix() {
        // literal control byte form
        let p = t("WpDcVyo5fP:\u{0001}tags\u{0001}product-0f483ff8ad7735e5e2dc0f44728d7bdb");
        assert_eq!(p.namespace, "WpDcVyo5fP");
        assert_eq!(p.key_type, "\\x01tags\\x01product");
    }

    #[test]
    fn preserves_already_escaped_tag_prefix() {
        let p = t("WpDcVyo5fP:\\x01tags\\x01product-0f483ff8ad7735e5e2dc0f44728d7bdb");
        assert_eq!(p.key_type, "\\x01tags\\x01product");
    }

    #[test]
    fn review_route_single_hash() {
        let p = t("WpDcVyo5fP:product-review-route-edc0e4a2cce33d9d8122d93bbe90a6f0-58f97500f374ea9a13ace9f52c7491b3");
        assert_eq!(p.key_type, "product-review-route");
    }

    #[test]
    fn no_colon_falls_back_to_prefix() {
        let p = t("plainkeywithoutcolon12345");
        assert_eq!(p.namespace, "plainkeywi");
        assert_eq!(p.key_type, "(no-type)");
    }

    #[test]
    fn short_hex_word_is_not_treated_as_hash() {
        // "cafe" is hex but too short — must be kept as a real type token.
        let p = t("WpDcVyo5fP:cafe");
        assert_eq!(p.key_type, "cafe");
    }

    #[test]
    fn body_that_is_only_a_hash() {
        let p = t("WpDcVyo5fP:0f483ff8ad7735e5e2dc0f44728d7bdb");
        assert_eq!(p.key_type, "(hash)");
    }
}

File: internal/grouping.go
package internal

import (
	"net/url"
	"strings"
)

const tagMarker = "\x01tags\x01"
const escapedTagMarker = "\\x01tags\\x01"

// ParsedKey is the result of parsing a single key.
type ParsedKey struct {
	Namespace string
	KeyType   string
}

// isHashSegment returns true if s looks like a hex hash segment (md5/sha-ish).
func isHashSegment(s string) bool {
	if len(s) < 16 {
		return false
	}
	for _, r := range s {
		if !isHexDigit(r) {
			return false
		}
	}
	return true
}

func isHexDigit(r rune) bool {
	return (r >= '0' && r <= '9') || (r >= 'a' && r <= 'f') || (r >= 'A' && r <= 'F')
}

// ParseKey splits a raw key into namespace + normalized type.
// namespaceLen is how many leading bytes of the key are treated as the
// namespace when there is no colon.
func ParseKey(key string, namespaceLen int) ParsedKey {
	// Render the 0x01 byte as the readable token "\x01".
	normalized := strings.ReplaceAll(key, tagMarker, escapedTagMarker)

	// Namespace = text before the first ':'. Fall back to first N chars.
	var namespace, body string
	if idx := strings.Index(normalized, ":"); idx >= 0 {
		namespace = normalized[:idx]
		body = normalized[idx+1:]
	} else {
		n := namespaceLen
		if n > len(normalized) {
			n = len(normalized)
		}
		namespace = normalized[:n]
		body = ""
	}

	keyType := normalizeType(body)
	return ParsedKey{Namespace: namespace, KeyType: keyType}
}

// normalizeType strips trailing hash segments from the key body to derive a stable type name.
func normalizeType(body string) string {
	if body == "" {
		return "(no-type)"
	}

	// Preserve the \x01tags\x01 prefix verbatim, normalize only the remainder.
	var prefix, rest string
	if strings.HasPrefix(body, escapedTagMarker) {
		prefix = escapedTagMarker
		rest = body[len(escapedTagMarker):]
	} else {
		prefix = ""
		rest = body
	}

	segments := strings.Split(rest, "-")

	// Pop trailing hash-looking segments, always keeping at least the first.
	for len(segments) > 1 && isHashSegment(segments[len(segments)-1]) {
		segments = segments[:len(segments)-1]
	}

	// Edge case: body that is *only* a hash.
	if len(segments) == 1 && isHashSegment(segments[0]) {
		return prefix + "(hash)"
	}

	return prefix + strings.Join(segments, "-")
}

// IsTagType returns true if the type name represents a Symfony tag set.
func IsTagType(name string) bool {
	return strings.Contains(name, escapedTagMarker)
}

// DisplayType humanizes a type name for the TUI. The Symfony tag-index set
// is shown with a visible badge so it stands out regardless of terminal emoji support.
func DisplayType(name string) string {
	if strings.Contains(name, escapedTagMarker) {
		tag := strings.ReplaceAll(name, escapedTagMarker, "")
		if unescaped, err := url.QueryUnescape(tag); err == nil && unescaped != tag {
			tag = unescaped
		}
		return "🏷 " + tag + "  [tag]"
	}
	if unescaped, err := url.QueryUnescape(name); err == nil && unescaped != name {
		return unescaped
	}
	return name
}

// DisplayNamespace humanizes a namespace for the TUI.
func DisplayNamespace(ns string) string {
	if unescaped, err := url.QueryUnescape(ns); err == nil && unescaped != ns {
		return unescaped
	}
	return ns
}

// DisplayKey humanizes a full key for the TUI (biggest-keys view).
func DisplayKey(key string) string {
	idx := strings.Index(key, escapedTagMarker)
	if idx >= 0 {
		before := key[:idx]
		tag := key[idx+len(escapedTagMarker):]
		
		if tag == "" {
			colonIdx := strings.Index(before, ":")
			if colonIdx >= 0 {
				tagName := before[colonIdx+1:]
				return before[:colonIdx+1] + "🏷 tags » " + tagName
			}
			return "🏷 tags » " + before
		}
		
		return before + "🏷 tags » " + tag
	}
	if unescaped, err := url.QueryUnescape(key); err == nil && unescaped != key {
		return unescaped
	}
	return key
}

// EscapeKey renders control bytes in a key name readably.
func EscapeKey(bytes []byte) string {
	return strings.ReplaceAll(string(bytes), tagMarker, escapedTagMarker)
}

// UnescapeKey restores control bytes from escaped form.
func UnescapeKey(key string) []byte {
	return []byte(strings.ReplaceAll(key, escapedTagMarker, tagMarker))
}

// MatchPatternFor builds the SCAN MATCH pattern for a (namespace, type).
func MatchPatternFor(namespace, keyType string) string {
	body := strings.ReplaceAll(keyType, escapedTagMarker, tagMarker)
	if body == "(no-type)" {
		return namespace + ":*"
	}
	if strings.HasSuffix(body, "(hash)") {
		stripped := strings.TrimSuffix(body, "(hash)")
		return namespace + ":" + stripped + "*"
	}
	return namespace + ":" + body + "*"
}
